### ISSUE TITLE
Websocket module: notifications through websocket

### ISSUE THREAD

**Issue by fgalan** [telefonicaid/fiware-orion#1669]
(Parent issue: #1181)

We could consider a new kind of subscriptions that instead of sending notifications through HTTP REST use the websocket used to create the subscription in order to send notifications back. A new URL schema (e.g. ws://) could be used so Orion can clearly identify them.

Probably this kind of subscriptions doesn’t need to be persisted in DB (it would live only in CB memory) given that once the websocket is closed by the client, then the subscription is immediately removed.

[ See issue #1943 for details about notifications in Active-Active environment ]

**Comment by fgalan** [comment]
After today discussions on skype, more info on how implement this. 

Let's consider that ws module keeps a table relating subIds (as string) and ws_hdls (websocket handles, in the structure used by the module, @fortizc I understand that your control that part). Table implementation details are out of the scope of this description (although probably a `std::map` could be used). Let's name this tables subsWs for convenience in this description (please, use a better name at implementation stage if you find so).

At CB startup time, the table starts empty. Let's consider the different flows:

* Subscription done through websocket. The ws module analyzes if the subscription is of ws nature looking to the reference for the `ws://` string. In that case, it processes the message as usual but looks the subscription response before sending it back to the client to get the subId associated to the subscription. A new entry is added to subsWs linking the subscriptionId to the ws handler (which is kept opened).

* Update context triggering the subscription. The CB processes this update as usual (no matter if it arrived on the REST API or in the WS API). In some point of mongoBackend the notification is triggered and Notifier::sendNotifyContextRequest() is invoked. This function will be modified to look at the reference string:
 * If reference starts with `http://` or `https://` then perform regular process.
 * If reference starts with `ws://` then a callback to ws module is done, including the subId and the notification content. Let's name this function sendNotifyContextRequestWs() by the moment (please, use a better name at implementation stage if you find so). This functions looks for subId in the subsWs table to get the corresponding ws_hdls. The ws module sends the notification using ws_hdl. More detail on sendNotifyContextRequestWs() signature in a separate comment.

* If the client closes the ws_hdl, the the ws module does a reverve lookup in the table to get the associated subId (if any) and in that case does a unsubscribe for subId on the lower layer and remove the entry from the table.

**Comment by fgalan** [comment]
Draft for sendNotifyContextRequestWs() signature:

```
bool sendNotifyContextRequestWs(std::string subId, std::map<string, string> headers, std::string data);
```

The JSON send back to the client to the ws handler associated to the subscription would be as follows:

```
{
   "subscriptionId": <subId>,
   "headers": <a keymap rendered from the headers map>,
   "data": <data string, it is assumed that the caller in Notifier module would prepare this to be a valid JSON>
}
```

The return value is `bool` to mean sucess or failure but we can set it to `void` if we find that at the end is not needed or to `int` if we want to have more granularity about the cause of the problems.

**Comment by fgalan** [comment]
Suggested work split:

* @fortizc: all the needed modifications in the ws module, including sendNotifyContextRequestWs() implementation.
* @fgalan: client part of the sendNotifyContextRequestWs() in the Notifier module.

**Comment by fgalan** [comment]
**Important remarks:** in the above descriptions, "ws://" refers to an exact text (of 5 chars), not to a prefix as it could be "http://".

**Comment by fgalan** [comment]
Implemented in PR #2004
